lib/b3fs/src/bucket/dir/reader.rs

```diff
@@ -1,35 +1,347 @@
 
 pub struct B3Dir {
     /// The number of entries in this directory.
     num_entries: u16,
     /// The actual header file.
     file: Arc<fs::File>,
+    bloom_filter: OnceCell<BloomFilter>,
+    hashtree: OnceCell<HashTree<'static>>,
+    position_start_hashes: u64,
+    position_start_bloom_filter: u64,
+    position_start_entries: u64,
+    position_start_phf_table: u64,
+    phf_table: OnceCell<HasherState>,
+}
+
+#[macro_use]
+macro_rules! check_eof {
+    ($flag:expr) => {
+        if let Err(e) = $flag {
+            if e.kind() == std::io::ErrorKind::UnexpectedEof {
+                return Ok(None);
+            } else {
+                return Err(errors::ReadError::from(e));
+            }
+        }
+    };
 }
 
 impl B3Dir {
     pub(crate) fn new(num_entries: u32, file: Arc<fs::File>) -> Self {
-        debug_assert!(num_entries <= (u16::MAX as u32));
+        debug_assert!(num_entries <= u16::MAX as u32);
+        let position_start_hashes = 4 + 4 as u64;
+        let position_start_bloom_filter =
+            position_start_hashes + (num_entries as usize * 32) as u64;
+        let length_bloom_filter = 4 + num_entries as usize;
+        let position_start_phf_table = position_start_bloom_filter + length_bloom_filter as u64;
+        let key_phf_table = 8;
+        let disps_len = (num_entries as f64 / 5.0).ceil() as usize * 4;
+        let entries_len = num_entries as usize * 4;
+        let length_phf_table = key_phf_table + disps_len + entries_len;
+        let position_start_entries =
+            position_start_bloom_filter + length_bloom_filter as u64 + length_phf_table as u64;
         Self {
             num_entries: num_entries as u16,
             file,
+            bloom_filter: OnceCell::new(),
+            hashtree: OnceCell::new(),
+            position_start_hashes,
+            position_start_bloom_filter,
+            position_start_entries,
+            position_start_phf_table,
+            phf_table: OnceCell::new(),
+        }
+    }
+
+    pub async fn hashtree(&mut self) -> Result<&HashTree<'static>, errors::ReadError> {
+        self.hashtree
+            .get_or_try_init(|| async {
+                let mut buffer = Vec::with_capacity(self.num_entries as usize * 32);
+                let mut file_reader = self.position_file(self.position_start_hashes).await?;
+                file_reader.read_exact(&mut buffer).await?;
+                let boxed_slice = buffer.into_boxed_slice();
+                let static_slice: &'static [u8] = Box::leak(boxed_slice);
+                HashTree::try_from(static_slice).map_err(|e| errors::ReadError::InvalidHashtree(e))
+            })
+            .await
+    }
+
+    pub async fn get_entry<'a>(
+        &'a mut self,
+        name: &'a [u8],
+    ) -> Result<Option<BorrowedEntry<'a>>, errors::ReadError> {
+        let bloom_filter = self.get_bloom_filter().await?;
+        if bloom_filter.contains(name) {
+            let phf_table = self.get_phf_table().await?;
```

Comment:
The point of having the PHF table in this use case is to minimize our IO, loading the entire table in memory is not the move.

Instead we could try something like:

```rs
let key = self.get_phf_key(); // cached
let hash = hash(entry, key);

let bucket = (hashes.g as usize) % phf_table.disps.len();
let bucket_offset = self.position_start_phf_table + 8 + bucket * 4;

let mut bucket_keys = [0; 4];
self.read_at_position(bucket_offset, &mut bucket_keys);

let d1 = u16::from_le_bytes(*array_ref!(bucket_keys, 0, 4));
let d2 = u16::from_le_bytes(*array_ref!(bucket_keys, 4, 4));

let idx = displace(hashes.f1, hashes.f2, d1 as u32, d2 as u32) as usize % self.num_entries;

let entry_rel_offset = [0; 4];
let map_offset = self.phf_map_offset + idx * 4;
self.read_at_position(bucket_offset, &mut entry_rel_offset);
let entry_rel_offset = u32::from_le_bytes(entry_rel_offset);

// self.entries_offset = after the hash tree ends
let entry_offset = self.entries_offset + entry_rel_offset;

return self.read_entry_at_offset(name, entry_offset).await;
```

Reply:
I see. I tried this but it is still not doing correctly. BTW... this is similar that the code i did. I was avoiding IO using phf table already. 

I am starting thinking that i am not handling properly the positions when i am pushing to the PHFTable, where you need to send the position.

Can you check that @qti3e ?